keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,364 @@
+fine-grained access control that enables Kubelet to set or update resources
+allocated to a Pod.
+
+#### Container Resize Policy
+
+To provide fine-grained user control, PodSpec.Containers is extended with
+ResizePolicy map (new object) for each resource type (CPU, memory):
+* NoRestart - the default value; resize Container without restarting it,
+* RestartContainer - restart the Container in-place to apply new resource
+  values. (e.g. Java process needs to change its Xmx flag)
+
+By using ResizePolicy, user can mark Containers as safe (or unsafe) for
+in-place resource update. Kubelet uses it to determine the required action.
+
+Setting the flag to separately control CPU & memory is due to an observation
+that usually CPU can be added/removed without much problem whereas changes to
+available memory are more probable to require restarts.
+
+If more than one resource type with different policies are updated, then
+RestartContainer policy takes precedence over NoRestart policy.
+
+Additionally, if RestartPolicy is 'Never', ResizePolicy should be set to
+NoRestart in order to pass validation.
+
+#### CRI Changes
+
+Kubelet calls UpdateContainerResources CRI API which currently takes
+*runtimeapi.LinuxContainerResources* parameter that works for Docker and Kata,
+but not for Windows. This parameter changes to *runtimeapi.ContainerResources*,
+that is runtime agnostic, and will contain platform-specific information.
+
+### Kubelet and API Server Interaction
+
+When a new Pod is created, Scheduler is responsible for selecting a suitable
+Node that accommodates the Pod.
+
+For a newly created Pod, Spec.Containers[i].ResourcesAllocated must match
+Spec.Containers[i].Resources.Requests. When Kubelet admits a new Pod, values in
+Spec.Containers[i].ResourcesAllocated are used to determine if there is enough
+room to admit the Pod. Kubelet does not set Pod's ResourcesAllocated after
+admitting a new Pod.
+
+When a Pod resize is requested, Kubelet attempts to update the resources
+allocated to the Pod and its Containers. Kubelet first checks if the new
+desired resources can fit the Node allocable resources by computing the sum of
+resources allocated (Pod.Spec.Containers[i].ResourcesAllocated) for all Pods in
+the Node, except the Pod being resized. For the Pod being resized, it adds the
+new desired resources (i.e Spec.Containers[i].Resources.Requests) to the sum.
+* If new desired resources fit, Kubelet accepts the resize by updating
+  Pod.Spec.Containers[i].ResourcesAllocated via pods/resourceallocation
+  subresource, and then proceeds to invoke UpdateContainerResources CRI API
+  to update the Container resource limits. Once all Containers are successfully
+  updated, it updates Pod.Status.ContainerStatuses[i].Resources to reflect the
+  new resource values.
+* If new desired resources don't fit, Kubelet rejects the resize, and no
+  further action is taken.
+  - Kubelet retries the Pod resize at a later time.
```

Comment:
Do you think we could have some condition set by kubelet upon the first failure (and cleared once the resize takes place)?
I'd imagine this would be slightly more friendly towards the initiating actors to detect that the request has failed. It should also be more resilient than waiting for timeout which seems race-prone.

Reply:
To keep things simple for Kubelet, we removed the PodCondition after introducing ResourcesAllocated in PodSpec since we expect ResourcesAllocated to be updated fairly quickly for accepted Pod resize (latency of watch), and a small timeout at the initiating actor becomes a feasible option.

If I understood your point correctly, the race is with VPA deciding to evict at the same time Kubelet accepts the resize, updates ResourcesAllocated, and proceeds to update the container limits, but VPA evicts before it sees the ResourcesAllocated update. I think this race would exist even if we added the PodCondition.

Reply:
+1 on what @vinaykul said above. We discussed this at SIG Node intensively and decided the current proposal for the simplicity.

Reply:
@kgolab I hope the above answered your question. If not, please reopen this issue.